### ISSUE TITLE
[postcss-custom-media] : replacement focussed on `importFrom` / `exportTo`

### ISSUE THREAD

**Issue by romainmenke** [csstools/postcss-plugins#531]
see : https://github.com/csstools/postcss-plugins/pull/527

- `importFrom` and `exportTo` have been removed from this plugin.
- a new plugin needs to be created to replace it.
- this new plugin would only do `importFrom` / `exportTo`

_Although we have the design tokens plugin it is not a drop in replacement and has a much smaller scope. Providing a drop in replacement will cause less issues for users._

**Comment by TommySorensen** [comment]
@romainmenke Is there a temporary solution to get `exportTo` working? I have a project where i need the same media queries in JS as in the CSS?

**Comment by romainmenke** [comment]
@TommySorensen the current advised way to do this is with [`@csstools/postcss-extract`](https://www.npmjs.com/package/@csstools/postcss-extract):

```js
const postcss = require('postcss');
const postcssCustomMedia = require('postcss-custom-media');
const postcssExtract = require('@csstools/postcss-extract');

postcss([
	postcssExtract({
		queries: {
			customMedia: 'atrule[name="custom-media"]'
		},
		results: (results) => {
			const formatted = {}

			results.customMedia.forEach((atRule) => {
				const splitIndex = atRule.params.indexOf(' '); // split after the custom media name
				formatted[atRule.params.slice(0, splitIndex)] = atRule.params.slice(splitIndex+1).trim();
			});

			console.log({
				customMedia: formatted
			});
		}
	}),
	postcssCustomMedia(/* pluginOptions */)
]).process(YOUR_CSS /*, processOptions */);
```

**Comment by k2snowman69** [comment]
Similar question for `importFrom`? I have a file where I defined all my custom media in one place:
```css
/* ./src/shared/custom-media.css */
@custom-media --media-1 (max-width: 799px);
@custom-media --media-2 (min-width: 800px) and (max-width: 1199px);
```

Then I used importFrom to effectively prefix these shared custom-media definitions to all css files being parsed:
```js
    postcssCustomMedia({
      importFrom: ["./src/shared/custom-media.css" ],
      preserve: false,
    }),
```

Finally in my css files I'd utilize the custom media query:
```css
@media (--media-1) {
  color: green;
}
```

However without importFrom, when my css file gets processed it doesn't know what `--media-1` is and thus doesn't process the token.

So far I've tried using `@import "./src/shared/custom-media.css"` which just appended the file, not resulting in processing the value.

I did see [this comment](https://github.com/csstools/postcss-plugins/issues/725#issuecomment-1322436540) which made me wonder if moving to design-tokens is the recommended migration path here? Any advice would be appreciated!

**Comment by k2snowman69** [comment]
That worked flawlessly! For anyone else coming here, all I needed to do is:

1. Whatever file was using definitions in `./src/shared/custom-media.css`, make sure to add `@import "./src/shared/custom-media.css"` at the top of the file.
2. Install postcss-import exactly as documented in the readme without any configuration

And now it successfully resolves the custom-media query! Thanks @romainmenke!

**Comment by romainmenke** [comment]
Hi everyone 👋,

The drop-in replacements have been created but we will not be releasing this.
We have decided against taking on maintenance for these plugins.

Anyone is free to take the source code for these replacement plugins and release them as their own!

https://github.com/csstools/postcss-plugins/pull/734

**Comment by kenvatian** [comment]
> That worked flawlessly! For anyone else coming here, all I needed to do is:
> 
> 1. Whatever file was using definitions in `./src/shared/custom-media.css`, make sure to add `@import "./src/shared/custom-media.css"` at the top of the file.
> 2. Install postcss-import exactly as documented in the readme without any configuration
> 
> And now it successfully resolves the custom-media query! Thanks @romainmenke!

**tl;dr**
Is there a way that we can `@import` once and use everywhere?

**gory details**
My setup is:

```
// file: postcss.config.js
module.exports = {
  entry: {
    styles: "./app/assets/stylesheets/styles.css",
  },
  plugins: [
    require("postcss-preset-env")({
      stage: 3,
      features: {
        'custom-media-queries': true,
      },
    }),
  ],
}
```

```
/* file: app/assets/stylesheets/util/custom-media.css */
@custom-media --small-and-above (min-width: 640px);
```

```
/* file: app/assets/stylesheets/variables.css */
@import "./util/custom-media.css";
```

```
/* file: app/assets/stylesheets/buttons.css */

/* doesn't work */
@media(--small-and-above) {
  .button {
    padding: 10px;
  }
}

/* but this works */
@import "./util/custom-media.css";
@media(--small-and-above) {
  .button {
    padding: 10px;
  }
}
```

```
/* file: app/assets/stylesheets/styles.css */
@import "./variables.css";
@import "./buttons.css"; /* custom media didn't get compiled from variables.css */
```

**Comment by romainmenke** [comment]
@kenvatian Do you have multiple entry points for your CSS?

multiple entry points :

```js
// in app.js

@import 'component-a.css';
@import 'component-b.css';
```

single entry point :

```js
// in app.js

@import 'styles.css';
```

```css
/* in styles.css */

@import 'component-a.css';
@import 'component-b.css';
```

**Comment by kenvatian** [comment]
Yes, we have multiple entry points.

In the example above at https://github.com/csstools/postcss-plugins/issues/531#issuecomment-1353078002.

component-a.css contains custom media queries. But it seems those queries in component-a.css are not picked up in component-b.css.

Thanks.

P.S. I notice that preserve is false by default in postcss-custom-media, maybe I try setting preserve to true?

**Comment by romainmenke** [comment]
@kenvatian Thank you for that extra info.

Multiple entry points are processed by `Webpack / PostCSS Loader` as completely separate stylesheets.
They do not share any globals like `@custom-media`.

`postcss-custom-media` is a polyfill only for `@custom-media` and should not be used to work around how `Webpack / PostCSS Loader` works internally.

The best way is to write 10 `@import` statements if you have 10 entry points.
That way you are writing CSS that matches how `Webpack / PostCSS Loader` and browsers work without any magic.

Alternatively :
- you can create a PostCSS plugin to import globals automatically
- you can release your own version of the `importFrom` functionality based on what we prepared here : https://github.com/csstools/postcss-plugins/pull/734
